File: smartlib/notifications/notification_service.py
"""Notification dispatch service."""
from typing import Optional, List
from smartlib.notifications.models import Notification
from smartlib.notifications.repository import NotificationRepository
from smartlib.constants import NotificationType

class NotificationService:
    def __init__(self, repo: Optional[NotificationRepository] = None):
        self.repo = repo or NotificationRepository()

    def send_notification(
        self,
        user_id: int,
        title: str,
        message: str,
        notification_type: str = NotificationType.GENERAL_ANNOUNCEMENT.value
    ) -> Notification:
        n = Notification(
            user_id=user_id,
            title=title,
            message=message,
            type=notification_type,
            is_read=False
        )
        return self.repo.create(n)

    def get_user_notifications(self, user_id: int, unread_only: bool = False) -> List[Notification]:
        return self.repo.list_by_user(user_id=user_id, unread_only=unread_only)

    def mark_read(self, notification_id: int, user_id: int) -> None:
        self.repo.mark_as_read(notification_id, user_id)

File: src/main/java/com/library/erp/service/NotificationService.java
package com.library.erp.service;

import com.library.erp.entity.Notification;
import com.library.erp.entity.enums.NotificationType;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface NotificationService {
    Notification sendNotification(Long userId, String title, String message, NotificationType type, String linkUrl);
    Page<Notification> getUserNotifications(Long userId, Pageable pageable);
    List<Notification> getRecentUnread(Long userId);
    long countUnreadNotifications(Long userId);
    void markAsRead(Long notificationId, Long userId);
    void markAllAsRead(Long userId);
}
